### PR TITLE
MLIBZ-2396: bugfix for memory consumption

### PR DESCRIPTION
**Pull request by heyzooi**
#### Description

Bugfix for memory consumption during sync/find/pull

#### Changes

- Discarding Sync/Wait result for auto-pagination requests

#### Tests

- Unit Test added and compare memory before and after a `pull()`

### CODE REVIEW

File: Kinvey/Kinvey/PullOperation.swift [hunk 1]
```diff
@@ -19,6 +19,7 @@ internal class PullOperation<T: Persistable>: FindOperation<T> where T: NSObject
         validationStrategy: ValidationStrategy?,
         cache: AnyCache<T>?,
         options: Options?,
+        mustSetRequestResult: Bool = true,
```
Comment (vinaygahlawat) [inline Kinvey/Kinvey/PullOperation.swift]: Under what circumstances is `mustSetRequestResult` set to false?

File: Kinvey/KinveyTests/SyncStoreTests.swift [hunk 2]
```diff
@@ -2785,4 +2785,67 @@ class SyncStoreTests: StoreTestCase {
...
+            switch $0 {
+            case .success(let results):
+                XCTAssertEqual(results.count, Int64(size))
+            case .failure(let error):
+                XCTFail(error.localizedDescription)
+            }
+            
+            if let startMemory = startMemory, let endMemory = reportMemory() {
+                let diffMemory = endMemory - startMemory
+                XCTAssertLessThan(diffMemory, 300_000_000)
```
Comment (vinaygahlawat) [inline Kinvey/KinveyTests/SyncStoreTests.swift]: Where does the comparison against 300,000,000 come from?

File: Kinvey/Kinvey/DataStore.swift [hunk 3]
```diff
@@ -9,28 +9,6 @@
 import Foundation
 import PromiseKit
 
-class DataStoreTypeTag: Hashable {
```
Comment (vinaygahlawat) [inline Kinvey/Kinvey/DataStore.swift]: Was the use of `DataStoreTypeTag` a cause of the memory leak?

File: Kinvey/Kinvey/Client.swift [hunk 4]
```diff
@@ -91,7 +90,17 @@ open class Client: Credential {
...
+        get {
+            return options?.timeout ?? Client.urlSessionConfiguration.timeoutIntervalForRequest
+        }
+        set {
+            if options == nil {
+                options = Options()
+            }
+            options?.timeout = newValue
+        }
+    }
```
[the remark this answers is not in the document]
Comment (heyzooi) [reply Kinvey/Kinvey/Client.swift]: none, i should create a separate PR for that. Doing that now.

File: Kinvey/KinveyTests/SyncStoreTests.swift (hunk 2 above)
Comment (heyzooi) [reply Kinvey/KinveyTests/SyncStoreTests.swift]: basically during my tests, the consumption was around `350 MB` without the fix and `250 MB` with the fix, that's why I pick `300 MB` as a threshold

File: Kinvey/Kinvey/DataStore.swift (hunk 3 above)
Comment (heyzooi) [reply Kinvey/Kinvey/DataStore.swift]: not really, again, i should create a separate PR for that. Doing that now.

File: Kinvey/Kinvey/PullOperation.swift (hunk 1 above)
Comment (heyzooi) [reply Kinvey/Kinvey/PullOperation.swift]: for auto pagination requests, if you set the result for each request you end up holding all instances in memory, that's why the memory consumption was going to the roof, so in this case we should `false` for this. This is all internal code, so no need to the user to set this.

**Comment by heyzooi** [comment]
@tejasranade and @vinaygahlawat, the unit tests are failing for now until we merge and rebase https://github.com/Kinvey/swift-sdk/pull/272

**Pull request merged**